File: examples/go/case03_test.go
package examplesgo

import (
	"errors"
	"testing"
)

type recordingBroker struct {
	events []string
	fail   bool
}

func (b *recordingBroker) Publish(event string) error {
	if b.fail {
		return errors.New("broker unavailable")
	}
	b.events = append(b.events, event)
	return nil
}

func TestOutboxClosesCommitGapButAllowsDuplicatePublish(t *testing.T) {
	weak := &OrderDB{}
	CommitBusinessStateOnly(weak) // Process stops before Publish is invoked.
	if !weak.Confirmed || weak.OutboxPending {
		t.Fatal("weak dual write did not expose the pre-publish crash gap")
	}
	rolledBack := &OrderDB{}
	if ConfirmWithOutbox(rolledBack, true) == nil || rolledBack.Confirmed || rolledBack.OutboxPending {
		t.Fatal("rolled-back transaction left a partial fact")
	}
	repaired, broker := &OrderDB{}, &recordingBroker{}
	if err := ConfirmWithOutbox(repaired, false); err != nil {
		t.Fatal(err)
	}
	if !repaired.Confirmed || !repaired.OutboxPending {
		t.Fatal("commit lost publication obligation")
	}
	if err := RelayPublishPending(repaired, broker); err != nil {
		t.Fatal(err)
	}
	// Relay crashes before MarkPublicationComplete; pending work is retried.
	if err := RelayPublishPending(repaired, broker); err != nil {
		t.Fatal(err)
	}
	if len(broker.events) != 2 {
		t.Fatal("relay crash did not expose duplicate publication")
	}
	MarkPublicationComplete(repaired)
	if err := RelayPublishPending(repaired, broker); err != nil {
		t.Fatal(err)
	}
	if len(broker.events) != 2 {
		t.Fatal("sent record was published again")
	}
}

File: examples/cpp/case03_test.cpp
#include <cassert>
#include <stdexcept>
#include <vector>
#include "case03.cpp"

int main() {
    Case03OrderDb weak;
    case03_commit_business_state_only(weak); // Process stops before publish is invoked.
    assert(weak.confirmed && !weak.outbox_pending);
    Case03OrderDb rolled_back;
    try {
        case03_confirm_with_outbox(rolled_back, true);
        assert(false);
    } catch (const std::runtime_error&) {
        assert(!rolled_back.confirmed && !rolled_back.outbox_pending);
    }
    Case03OrderDb repaired;
    case03_confirm_with_outbox(repaired, false);
    assert(repaired.confirmed && repaired.outbox_pending);
    std::vector<std::string> events;
    auto publish = [&](const std::string& event) { events.push_back(event); };
    case03_relay_publish_pending(repaired, publish);
    case03_relay_publish_pending(repaired, publish); // Crash before marking SENT.
    assert(events.size() == 2);
    case03_mark_publication_complete(repaired);
    case03_relay_publish_pending(repaired, publish);
    assert(events.size() == 2);
}
